bftengine/src/bftengine/CryptoManager.cpp

```diff
@@ -0,0 +1,245 @@
+    }
+  }
+}
+
+void CryptoManager::onPrivateKeyExchange(const std::string& secretKey,
+                                         const std::string& verificationKey,
+                                         const SeqNum& sn) {
+  LOG_INFO(logger(), "Private key exchange:" << KVLOG(sn, verificationKey));
+  auto sys_wrapper = create(sn);
+  sys_wrapper->cryptosys_->updateKeys(secretKey, verificationKey);
+  sys_wrapper->init();
+}
+
+void CryptoManager::onPublicKeyExchange(const std::string& verificationKey,
+                                        const std::uint16_t& signerIndex,
+                                        const SeqNum& sn) {
+  LOG_INFO(logger(), "Public key exchange:" << KVLOG(sn, signerIndex, verificationKey));
+  auto sys = create(sn);
+  sys->cryptosys_->updateVerificationKey(verificationKey, signerIndex);
+  sys->init();
+}
+
+void CryptoManager::onCheckpoint(uint64_t newCheckpoint) {
+  std::lock_guard<std::mutex> guard(mutex_);
+  std::vector<CheckpointNum> checkpointsToRemove;
+  auto checkpointOfCurrentCryptoSystem = getCheckpointOfCryptosystemForSeq(newCheckpoint * checkpointWindowSize);
+
+  std::experimental::erase_if(
+      cryptoSystems_, [this, checkpointOfCurrentCryptoSystem, newCheckpoint](const auto& checkpointToSystem) {
+        auto checkpoint = checkpointToSystem.first;
+        if (checkpoint < checkpointOfCurrentCryptoSystem) {
+          LOG_INFO(logger(), "Removing stale cryptosystem " << KVLOG(checkpoint, newCheckpoint, cryptoSystems_.size()));
+          return true;
+        }
+        return false;
+      });
+  assertMapSizeValid();
+}
+
+std::shared_ptr<EdDSAMultisigSigner> CryptoManager::getSigner(SeqNum seq) const {
+  auto signer = get(seq)->thresholdSigner_;
+  return std::reinterpret_pointer_cast<EdDSAMultisigSigner>(signer);
+}
+
+std::shared_ptr<EdDSAMultisigVerifier> CryptoManager::getMultisigVerifier(SeqNum seq) const {
+  auto verifier = get(seq)->thresholdVerifierForOptimisticCommit_;
+  return std::reinterpret_pointer_cast<EdDSAMultisigVerifier>(verifier);
+}
+
+std::array<std::pair<SeqNum, std::shared_ptr<EdDSAMultisigVerifier>>, 2> CryptoManager::getLatestVerifiers() const {
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto riter = checkpointToSystem().rbegin();
+  std::array<std::pair<SeqNum, std::shared_ptr<EdDSAMultisigVerifier>>, 2> result;
+  result[0] = {
+      riter->first,
+      std::reinterpret_pointer_cast<EdDSAMultisigVerifier>(riter->second->thresholdVerifierForOptimisticCommit_)};
+  riter++;
+  if (riter != checkpointToSystem().rend() && riter->second != nullptr) {
+    result[1] = {
+        riter->first,
+        std::reinterpret_pointer_cast<EdDSAMultisigVerifier>(riter->second->thresholdVerifierForOptimisticCommit_)};
+  }
+  return result;
+}
+
+std::array<std::shared_ptr<EdDSAMultisigSigner>, 2> CryptoManager::getLatestSigners() const {
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto riter = checkpointToSystem().rbegin();
```

Comment:
do you need to check if riter is valid?

Reply:
checkpointToSystem ensures that the size is valid, the mutex ensures that this is the only access - the iterator is expected to be valid.
As for checking that the cryptosystem is not null, I can add this check.